File: tests/r_parity/10_honest_did.py
"""StatsPAI Honest-DiD parity (Python side) -- Module 10.

Runs the **native** ``sp.honest_did(method="smoothness")`` fixed-length
confidence interval (the Rambachan-Roth FLCI under ``Delta^SD(M)``, solved
in Python by ``statspai.did._flci``) on a hand-crafted event study, at five
values of ``M``, and compares it with R ``HonestDiD`` on the same inputs.

Until 1.31 this module called ``backend="honestdid"``, which hands the
computation to the R package itself: the row then checked the wrapper, not
an independent Python algorithm (comparing an R bridge with R is circular).
It now exercises the native solver, and the R side supplies three
references so each gap has a named mechanism:

``ci_{lower,upper}_M_*``  (headline for ``M > 0``)
    ``HonestDiD::createSensitivityResults(method = "FLCI")`` with its
    internal ``.qfoldednormal`` replaced by the exact folded-normal
    quantile -- the like-for-like comparison. The Stata ``honestdid`` port
    joins these rows too.
``shipped_ci_{lower,upper}_M_*``
    HonestDiD as shipped. It evaluates the folded-normal quantile by
    simulation (1e6 draws, seed 0), so it sits ~3e-4 from the exact
    quantile: a documented difference in the computed quantity, displayed
    but not a parity row.
``analytic_ci_{lower,upper}_M_0``  (headline)
    At ``M = 0`` the violation is exactly linear, the worst-case bias is
    zero and the FLCI centre is the GLS linear extrapolation
    ``beta_post[0] - s_hat`` (``s_hat`` = GLS slope of the pre-period
    coefficients on relative time), with half-length ``z_{0.975}`` times its
    standard deviation. R computes this closed form with base linear
    algebra, independently of both implementations. HonestDiD's cone solver
    returns this interval to ~6e-6; the native solver to ~2e-9. The plain
    ``ci_*_M_0`` rows are therefore displayed but are not the headline.

The event study carries an explicit covariance (``event_study_vcov``):
the native FLCI deliberately refuses to fabricate a joint covariance from
standard errors alone, and the reference uses ``diag(se^2)``.

The closed-form ``breakdown_m`` primitive is independently pinned in
tests/external_parity/test_honest_did_paper_parity.py, and the FLCI
half-length against ``HonestDiD::findOptimalFLCI`` (exact quantile) at 1e-8
on a correlated covariance in tests/reference_parity/test_did_synth_R_parity.py.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp
from statspai.core.results import CausalResult

from _common import ParityRecord, write_results

MODULE = "10_honest_did"
M_GRID = [0.0, 0.05, 0.1, 0.2, 0.5]


def main() -> None:
    # Hand-crafted event study so the test is deterministic and the
    # R side can mirror the inputs without an intermediate CSV. Pre-period
    # relative_time -3, -2, -1; post-period 0, 1, 2.
    es = pd.DataFrame(
        {
            "relative_time": [-3, -2, -1, 0, 1, 2],
            "att": [0.01, -0.02, 0.0, 0.5, 0.4, 0.3],
            "se": [0.05, 0.05, 0.05, 0.10, 0.10, 0.10],
        }
    )
    vcov = pd.DataFrame(
        np.diag(es["se"].to_numpy() ** 2),
        index=es["relative_time"],
        columns=es["relative_time"],
    )
    res = CausalResult(
        method="ParityHonestDiDInput",
        estimand="ATT(0)",
        estimate=0.5,
        se=0.10,
        pvalue=0.0,
        ci=(0.30, 0.70),
        alpha=0.05,
        n_obs=1000,
        model_info={"event_study": es, "event_study_vcov": vcov},
    )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="breakdown_m_e0",
            estimate=float(sp.breakdown_m(res, e=0, method="smoothness", alpha=0.05)),
            n=int(res.n_obs),
        )
    ]

    table = sp.honest_did(res, e=0, m_grid=M_GRID, method="smoothness")
    if table.attrs.get("interval") != "flci":
        raise RuntimeError(
            "native honest_did fell back to the worst-case-bias interval; "
            "module 10 must exercise the native FLCI"
        )
    for _, row in table.iterrows():
        m = float(row["M"])
        prefixes = ["", "shipped_"] + (["analytic_"] if m == 0.0 else [])
        for prefix in prefixes:
            for side, col in (("lower", "ci_lower"), ("upper", "ci_upper")):
                rows.append(
                    ParityRecord(
                        module=MODULE,
                        side="py",
                        statistic=f"{prefix}ci_{side}_M_{m:g}",
                        estimate=float(row[col]),
                        n=int(res.n_obs),
                    )
                )

    write_results(
        MODULE,
        "py",
        rows,
        extra={
            "method": "smoothness",
            "backend": "native",
            "interval": "flci",
            "alpha": 0.05,
            "M_grid": M_GRID,
            "implementation": (
                "native Python FLCI (statspai.did._flci.flci_delta_sd) with "
                "the exact folded-normal quantile; no R call. The same "
                "native values are emitted under three statistic prefixes so "
                "they join the three R references (HonestDiD with the exact "
                "quantile, HonestDiD as shipped, closed form at M = 0)."
            ),
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/10_honest_did.R
# StatsPAI Honest-DiD parity (R side) -- Module 10.
#
# Mirrors the hand-crafted event study used by 10_honest_did.py, whose
# Python side runs the *native* StatsPAI FLCI (no R call). Three
# references are written so every gap has a named mechanism:
#
#   ci_*_M_*           HonestDiD::createSensitivityResults(method = "FLCI")
#                      with HonestDiD's internal .qfoldednormal replaced by
#                      the exact quantile (root of the folded-normal CDF).
#                      Like-for-like with the native solver; the headline
#                      rows for M > 0.
#   shipped_ci_*_M_*   HonestDiD as shipped. Its folded-normal quantile is
#                      simulated (1e6 draws, seed 0): ~3e-4 from exact.
#   analytic_ci_*_M_0  Closed form at M = 0, computed here with base linear
#                      algebra and no HonestDiD call: under Delta^SD(0) the
#                      violation is linear, the worst-case bias is zero, and
#                      the FLCI is the GLS linear extrapolation
#                      beta_post[1] - s_hat plus/minus z_{0.975} sd. It
#                      adjudicates the M = 0 gap between the two solvers.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(HonestDiD)
})

MODULE <- "10_honest_did"
M_GRID <- c(0.0, 0.05, 0.1, 0.2, 0.5)

#   pre  rel-time {-3,-2,-1} : att = (0.01, -0.02, 0.0), SE = 0.05
#   post rel-time { 0, 1, 2} : att = (0.5,   0.4,   0.3), SE = 0.10
betahat <- c(0.01, -0.02, 0.0,   0.5, 0.4, 0.3)
ses     <- c(0.05, 0.05,  0.05,  0.10, 0.10, 0.10)
sigma   <- diag(ses^2)

run_flci <- function() {
  # Suppress the cone-solver "may be inaccurate" warnings that come
  # from the small synthetic example.
  suppressWarnings(
    HonestDiD::createSensitivityResults(
      betahat = betahat,
      sigma   = sigma,
      numPrePeriods  = 3,
      numPostPeriods = 3,
      Mvec    = M_GRID,
      method  = "FLCI",
      alpha   = 0.05
    )
  )
}

rows <- list()
emit <- function(prefix, m, lb, ub) {
  rows[[length(rows) + 1L]] <<- parity_row(
    module = MODULE, statistic = sprintf("%sci_lower_M_%g", prefix, m),
    estimate = lb, n = 1000
  )
  rows[[length(rows) + 1L]] <<- parity_row(
    module = MODULE, statistic = sprintf("%sci_upper_M_%g", prefix, m),
    estimate = ub, n = 1000
  )
}

# (1) HonestDiD as shipped.
sens <- run_flci()
for (i in seq_along(M_GRID)) emit("shipped_", M_GRID[i], sens$lb[i], sens$ub[i])

# (2) HonestDiD with the exact folded-normal quantile.
exact_qfoldednormal <- function(p, mu = 0, sd = 1, numSims = 10^6, seed = 0) {
  vapply(mu, function(m) {
    m <- abs(m)
    stats::uniroot(
      function(q) stats::pnorm((q - m) / sd) - stats::pnorm((-q - m) / sd) - p,
      c(0, m + 20 * sd), tol = 1e-14
    )$root
  }, numeric(1))
}
shipped_qfoldednormal <- get(".qfoldednormal", envir = asNamespace("HonestDiD"))
assignInNamespace(".qfoldednormal", exact_qfoldednormal, ns = "HonestDiD")
sens_exact <- run_flci()
assignInNamespace(".qfoldednormal", shipped_qfoldednormal, ns = "HonestDiD")
for (i in seq_along(M_GRID)) emit("", M_GRID[i], sens_exact$lb[i], sens_exact$ub[i])

# (3) Closed form at M = 0: GLS slope through the omitted reference period.
t_pre   <- c(-3, -2, -1)
S_pre   <- sigma[1:3, 1:3]
Si      <- solve(S_pre)
s_hat   <- drop(solve(t(t_pre) %*% Si %*% t_pre) %*% t(t_pre) %*% Si %*% betahat[1:3])
var_s   <- drop(solve(t(t_pre) %*% Si %*% t_pre))
theta   <- betahat[4] - 1 * s_hat          # first post period sits at t = +1
sd_th   <- sqrt(sigma[4, 4] + 1^2 * var_s)  # pre/post independent here
z       <- stats::qnorm(0.975)
emit("analytic_", 0, theta - z * sd_th, theta + z * sd_th)

write_results(MODULE, rows,
              extra = list(method = "FLCI", alpha = 0.05,
                           references = c("shipped", "exact_quantile",
                                          "closed_form_M0")))
